### ISSUE TITLE
[Question] How difficult would it be to have dynamic location of the connection sockets on the node body?

### ISSUE THREAD

**Issue by MostHated** [alelievr/NodeGraphProcessor#60]
Hey there,
    I have been considering implementing a node graph into my project as it feels like it might make things a bit easier for me. I have just started looking at what's available but so far I like the feature set I am seeing with this one, especially the smaller connection relocation/relay node. The one thing I really need to make sure I can do is be able to implement dynamic placement of the actual connector sockets on the node body itself. In all the images I see so far from this framework, as well as the examples, they are only ever on the left and right sides of the node body. Are they able to be placed on the top and bottom of the node as well?

The only example I have to display what I mean is from an image of [another older node framework](https://github.com/nicloay/Node-Inspector):
![](https://i.imgur.com/bCCOfNw.png)

While the main features this framework are great, the actual connection socket being able to move position around the node body is the one thing that would end up making it worth using a node framework at all. If this sounds like something that could be fairly easily implemented then I will definitely give this framework a go.

I appreciate your time,
Thanks,
-MH

**Comment by alelievr** [comment]
Hello,

I've tried in the past to implement it but failed to find a design that pleased me. For reference, I wanted vertical ports just lilke VFX Graph has:
![image](https://user-images.githubusercontent.com/6877923/87303189-06f52100-c513-11ea-92e2-c0df37eb93c9.png)

That being said, it's not that hard to hack in the current project, I'd suggest that you take a look at this function: https://github.com/alelievr/NodeGraphProcessor/blob/master/Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseNodeView.cs#L255 where the visual port creation happens.
You could for example change the style of the port to absolute and set it's position to wherever you want in the node. The main issue then will be the edge connection direction which remains horizontal even if the port direction is set to vertical:
![image](https://user-images.githubusercontent.com/6877923/87303644-cb0e8b80-c513-11ea-8817-ea94d3e7348e.png)

**Comment by MostHated** [comment]
Thanks for the reply. I appreciate it. For me, personally, I think I would be ok with it as you showed above, it is just to make things a bit easier on myself. I have a waypoint system in which on each waypoint I have a list containing the current waypoints connections to other waypoints, so each node would of course be a waypoint and the connections would just be it's connections to others, to make it much easier to visualize and edit the connections vs finding them in the hierarchy and dragging them into a list one by one.

I was hoping to have the connection points on any of the 4 sides so that I could lay them out in a similar manner to how they appear in the actual scene( pic below), to make figuring out which to connect to which easier.
![](https://i.imgur.com/Sqenrns.png)

For example, in your pic, UpPortNode could represent the node WP_V_06 in my image below, then your Add Node's output would be node WP_V_13 in my image. So really, the fact that the line doesn't line up exactly as it should really is of minimal concern in this instance, just having it at the top would be a big help. Then depending on which node the connection originated from, that would be the direction the connection would flow. So starting the connection from the Add Nodes output and connecting it to UpPortNode, would then add the UpPortNode to the Add nodes list of connected waypoints.

![](https://i.imgur.com/4zeg8iP.png)

Then just depending on where the actual nodes are in comparison to each other would determine if the connection points should be in the top, bottom, etc, etc. I would imagine that this use case should work ok based on what you mentioned about it, I suppose I will just have to play around with it and see how it goes.

I do appreciate the information,
Thanks,
-MH

**Comment by alelievr** [comment]
Yeah, this sounds completely feasible. With the UIElement style system, you could even change the port position around the node to reflect the position in your waypoint system (moving them dynamically is also possible).

**Comment by MostHated** [comment]
My first test was not super successful, unfortunately, lol. I will jsut have to mess around with it more and see what I can come up with.


![](https://i.imgur.com/Ju4jBWi.png)
![](https://i.imgur.com/UQDuxlz.png)

**Comment by alelievr** [comment]
Ah yes, I didn't though about that, it's a very old issue I forgot to fix however the fix is simple, I pushed it on master: 71557d23db0254090d581af8fc460f03585a5311

Another limitation that may affect you is that you can't connect an input to another input, you can allow it by overriding the [GetCompatiblePorts](https://github.com/alelievr/NodeGraphProcessor/blob/master/Assets/com.alelievr.NodeGraphProcessor/Editor/Views/BaseGraphView.cs#L351) method.

**Comment by alelievr** [comment]
I changed the compute order in this commit: https://github.com/alelievr/NodeGraphProcessor/commit/c712b2861755da6489f258ae03253aa8c7dc4389, note that the new algorithm is this one: https://en.wikipedia.org/wiki/Depth-first_search

This shouldn't affect you, but just in case, the new algorithm set a compute order value of -2 when it detects cycles in the graph instead of logging an error.

If you encounter problems with the new algorithm, please let me know. Also you can revert to the old algorithm by changing
```CSharp
		public void UpdateComputeOrder(ComputeOrderType type = ComputeOrderType.DepthFirst)
```
to
```CSharp
		public void UpdateComputeOrder(ComputeOrderType type = ComputeOrderType.BreadthFirst)
```
